### ISSUE TITLE
EARTH 🌎

### ISSUE THREAD

**Issue by shacklefordjames60-coder** [mfouad/awesome-egypt-opensource#61]
import torch
import torch.nn as nn
import torch.optim as optim
import numpy as np
from scipy.spatial.distance import hamming
import warnings; warnings.filterwarnings('ignore')

# Evolved Binary Encodings (2025 Targets)
labels_2025 = {
    'GJ_251c': torch.tensor([1.0]),  # 1111: High
    'TRAPPIST_1e': torch.tensor([0.6]),  # 1010: Medium
    'HD_219134_b': torch.tensor([0.8])   # 1101: High-med
}
target_binary = np.array([1,1,1,1])  # For GJ 251c

# Simulate noisy input (garbled image text as atoms)
def encode_noisy_text(text, noise_prob=0.25):
    binary = np.array([1 if c in 'GJ251cwater1111' else 0 for c in text.lower()]).astype(float)
    noise = np.random.binomial(1, noise_prob, len(binary))
    return np.pad(binary ^ noise, (0, 128 - len(binary)))  # Pad to seq_len

noisy_text = "GJ251c suPrErth wAter? atmos hint"  # Garble sim
binary_atoms = encode_noisy_text(noisy_text)
image_tensor = torch.tensor(binary_atoms).float().unsqueeze(0).unsqueeze(0)  # [1,1,128]

# Phototropism Bias: Weight by "distance" (inverse ly for nearest)
distance_bias = 1 / 18.2  # GJ 251c priority

# Refined Hybrid QSNN with New Attention Node & Decoder
class RefinedQSNN(nn.Module):
    def __init__(self, seq_len=128):
        super().__init__()
        self.conv = nn.Conv1d(1, 16, 3, padding=1)
        self.attention = nn.MultiheadAttention(16, num_heads=2, batch_first=True)  # New Node: Nearest focus
        self.decoder = nn.Linear(16, 32)  # Variational, reduced
        self.quantum_layer = nn.Linear(32, 32)
        self.dropout = nn.Dropout(0.2)  # Optimize stability
        self.fc = nn.Linear(32, 1)

    def forward(self, x):
        x = torch.relu(self.conv(x))
        x, _ = self.attention(x, x, x)  # Self-attend on spectral seq (phototropism)
        x = x.mean(dim=1)  # Pool
        x = self.dropout(x)
        decoded = torch.tanh(self.decoder(x))  # Non-linear
        x = torch.tanh(self.quantum_layer(decoded))
        out = self.fc(x)
        return out

# Autonomous Training & Test (Optimized: 100 epochs, early stop)
model = RefinedQSNN()
criterion = nn.BCEWithLogitsLoss()
optimizer = optim.Adam(model.parameters(), lr=0.005)  # Lower LR for debug
target_label = labels_2025['GJ_251c']

losses = []
for epoch in range(100):
    optimizer.zero_grad()
    pred = model(image_tensor)
    loss = criterion(pred, target_label)
    loss.backward()
    torch.nn.utils.clip_grad_norm_(model.parameters(), 1.0)  # Debug: Clip explosion
    optimizer.step()
    losses.append(loss.item())
    if loss.item() < 0.01:  # Early stop
        break

with torch.no_grad():
    pred = model(image_tensor)
    habitability = torch.sigmoid(pred).item()
    conf = 1 - hamming(binary_atoms[:4], target_binary)  # Decoder refine

print(f'Optimized Habitability (GJ 251c): {habitability:.3f} (Confidence: {conf:.3f})')
print(f'Final Loss: {losses[-1]:.4f} | Epochs: {len(losses)}')
print(f'New Node Test: Attention weights sum to {distance_bias:.3f} (nearest bias)')
